tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+	 * grouping strategy for by-value types.
+	 */
+	Assert(false);
+#endif
+
+	get_key_arrow_fixed(column, row, 8, key, valid);
+}
+
+/*
+ * Implementation of bulk hashing specialized for a given key width.
+ */
+static pg_noinline uint32
+fill_offsets_arrow_fixed_8(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_8);
+}
+
+static pg_noinline uint32
+fill_offsets_arrow_fixed_4(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_4);
+}
+
+static pg_noinline uint32
+fill_offsets_arrow_fixed_2(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_2);
+}
+
+static pg_noinline uint32
+fill_offsets_scalar(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+					int key_column_index, uint32 next_unused_state_index, int start_row,
+					int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_scalar);
+}
+
+static void
+gp_hash_add_batch(GroupingPolicy *gp, DecompressBatchState *batch_state)
+{
+	GroupingPolicyHash *policy = (GroupingPolicyHash *) gp;
+
+	Assert(!policy->returning_results);
+
+	const uint64_t *restrict filter = batch_state->vector_qual_result;
+	const int n = batch_state->total_batch_rows;
+
+	/*
+	 * Initialize the array for storing the aggregate state offsets corresponding
+	 * to a given batch row.
+	 */
+	if ((size_t) n > policy->num_allocated_offsets)
+	{
+		policy->num_allocated_offsets = n;
+		policy->offsets = palloc(sizeof(policy->offsets[0]) * policy->num_allocated_offsets);
+	}
+	memset(policy->offsets, 0, n * sizeof(policy->offsets[0]));
+
+	/*
+	 * For the partial aggregation node, the grouping columns are always in the
+	 * output, so we don't have to separately look at the list of the grouping
+	 * columns.
+	 */
+	Assert(list_length(policy->output_grouping_columns) == 1);
```

Comment:
I'll keep it here for now, because in the following changes the hash grouping policy will choose a different hashing strategy here based on the type of grouping key (i.e. arithmetic/text/multicolumn).